File: Source/app/src/main/java/com/anshulvyas/csc780/grocerymanagr/HomeFragment.java
package com.anshulvyas.csc780.grocerymanagr;


import android.app.AlertDialog;
import android.app.Notification;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v4.app.ListFragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ListView;

import com.anshulvyas.csc780.grocerymanagr.Adapters.ProductAdapter;
import com.anshulvyas.csc780.grocerymanagr.Model.DBManager;

import org.joda.time.DateTime;
import org.joda.time.Days;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

/**
 * Hosts the listView of items that user wants to keep track of - with number of days before item expires
 */
public class HomeFragment extends Fragment {

    private FloatingActionButton mFAB;
    private List<Product> productList, filterProductList;
    private DBManager dbManager;
    private ListView productListView;
    private Context myContext;
    public List<Integer> notificationList;

    /**
     * Called when a fragment is first attached to its context.
     * @param context - to be attached to
     */
    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        myContext = context;
    }


    /**
     * Called to do initial creation of a fragment. (overridden)
     * @param savedInstanceState
     */
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        dbManager = new DBManager(getActivity());
    }

    /**
     * Called to have the fragment instantiate its user interface view.
     * @param inflater
     * @param container
     * @param savedInstanceState
     * @return view
     */
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {

        View view = inflater.inflate(R.layout.fragment_home, container, false);

        filterProductList = new ArrayList<>();

        productList = dbManager.getAllStockedProducts();
        if (productList.size() > 0) {
            Log.i("~!@#HOMEFRAGMENT", productList.get(0).toString());

            for (int i = 0; i < productList.size(); i++) {
                if (!productList.get(i).isShoppingCheck()) {
                    filterProductList.add(productList.get(i));
                }
            }

            notificationList = new ArrayList<>();
            notificationList.clear();

            for (int i = 0; i < filterProductList.size(); i++) {
                Product productObj = filterProductList.get(i);
                if(getLeftDays(productObj) <= 0) {
                    productObj.setExpired(true);
                    productObj.setStocked(false);
                    productObj.setConsumed(false);
                    dbManager.updateProduct(productObj);
                } else if(getLeftDays(productObj) == 1) {
                    if(productObj != null) {
                        notificationList.add(productObj.getProductId());
                    }
                }
            }

            createNotification();

            productListView = (ListView) view.findViewById(R.id.listView_home_product);
            final ProductAdapter productAdapter = new ProductAdapter(getActivity().getBaseContext(), R.layout.list_view_home,
                    filterProductList);
            productListView.setAdapter(productAdapter);
            productListView.setDivider(this.getResources().getDrawable(R.drawable.transparent));

            productAdapter.setNotifyOnChange(true);
            productAdapter.notifyDataSetChanged();

            /**
             * Calling a delete dialog box to confirm user's action before deleting the item from the Adapter.
             */
            productListView.setOnItemLongClickListener(new AdapterView.OnItemLongClickListener() {
                @Override
                public boolean onItemLongClick(AdapterView<?> parent, final View view, int position, long id) {

                    final AlertDialog.Builder deleteDialog = new AlertDialog.Builder(getActivity());
                    deleteDialog.setTitle("Delete?");
                    deleteDialog.setMessage("Are you sure you want to delete?");
                    final int viewPosition = position;

                    deleteDialog.setPositiveButton("OK", new AlertDialog.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            dbManager.deleteProduct(productAdapter.getItem(viewPosition));
                            Log.d("DEMO======>", "PRODUCT DELETED");
                            Snackbar.make(view, productAdapter.getItem(viewPosition).getProductName() + " deleted", Snackbar
                                    .LENGTH_LONG).show();

                            productAdapter.remove(productAdapter.getItem(viewPosition));


                            productAdapter.setNotifyOnChange(true);
                            productAdapter.notifyDataSetChanged();
                        }
                    });

                    deleteDialog.setNegativeButton("Cancel", new AlertDialog.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            dialog.dismiss();
                        }
                    });

                    deleteDialog.show();
                    return false;
                }
            });
        } else{
            Log.i("~!@#HOMEFRAGMENT", "list view not displayed");
        }

        /**
         * Instantiate Floating Action button
         */
        mFAB = (FloatingActionButton) view.findViewById(R.id.addFAB);
        mFAB.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
                    nextActivity();
                }
            });

        return view;
    }

    /**
     * Workaround for calling an intent from the fragment to an activity.
     */
    private void nextActivity() {
        Intent intent = new Intent(getActivity(), AddItemActivity.class);
        startActivity(intent);
    }

    /**
     * Get number of days till expiry. calculates the days between the currentDate an the selected expiry date.
     * @param productObj Product object
     * @return number of days
     */
    public int getLeftDays(Product productObj) {
        Calendar now = Calendar.getInstance();

        SimpleDateFormat format = new SimpleDateFormat("MM/dd/yyyy");
        Calendar expiry = Calendar.getInstance();
        try {
            expiry.setTime(format.parse(productObj.getExpiryDate()));
        } catch (ParseException e) {
            e.printStackTrace();
        }

        DateTime dateTimeNow = new DateTime(now.getTime());
        DateTime dateTimeExp = new DateTime(expiry.getTime());

        Days days = Days.daysBetween(dateTimeNow, dateTimeExp);

        return days.getDays() + 1;
    }

    /**
     * Creates notification when called!
     */
    public void createNotification() {
        // Prepare intent which is triggered if the
        // notification is selected
        Intent intent = new Intent(myContext, HomeActivity.class);
        Integer[] simpleArray = new Integer[notificationList.size()];
        notificationList.toArray(simpleArray);

        intent.setFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP);
        intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
        intent.putExtra( "List", simpleArray);
        PendingIntent pIntent = PendingIntent.getActivity(myContext, (int) System.currentTimeMillis(), intent, 0);


        Notification notification = new Notification.Builder(myContext)
                .setContentTitle("Stocked!")
                .setContentText(String.valueOf(notificationList.size()) + " Item(s) about to expire")
                .setSmallIcon(R.mipmap.ic_launcher)
                .setContentIntent(pIntent)
                .build();
        NotificationManager notificationManager = (NotificationManager) myContext.getSystemService(Context.NOTIFICATION_SERVICE);

        // hide the notification after its selected
        notification.flags |= Notification.FLAG_AUTO_CANCEL;

        notificationManager.notify(0, notification);
    }

//
//    @Override
//    public void onDismiss(DialogInterface dialog) {
//        final ProductAdapter productAdapter = new ProductAdapter(getActivity().getBaseContext(), R.layout.list_view_home,
//                filterProductList);
//        productListView.setAdapter(productAdapter);
//        productListView.setDivider(this.getResources().getDrawable(R.drawable.transparent));
//        productListView.setDividerHeight(20);
//
//        productAdapter.setNotifyOnChange(true);
//        productAdapter.notifyDataSetChanged();
//    }

}

File: app/src/main/java/io/github/anshu7vyas/stocked/ui/home/HomeFragment.kt
package io.github.anshu7vyas.stocked.ui.home

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ListView
import androidx.appcompat.app.AlertDialog
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.android.material.snackbar.Snackbar
import dagger.hilt.android.AndroidEntryPoint
import io.github.anshu7vyas.stocked.R
import io.github.anshu7vyas.stocked.ui.additem.AddItemActivity
import kotlinx.coroutines.launch

/** Pantry tab: stocked items with days-until-expiry, long-press for actions. */
@AndroidEntryPoint
class HomeFragment : Fragment() {

    private val viewModel: HomeViewModel by viewModels()

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View {
        val view = inflater.inflate(R.layout.fragment_home, container, false)

        val adapter = StockedItemAdapter(requireContext())
        val listView = view.findViewById<ListView>(R.id.listView_home_product)
        listView.adapter = adapter
        listView.divider = ContextCompat.getDrawable(requireContext(), R.drawable.transparent)

        listView.setOnItemLongClickListener { _, rowView, position, _ ->
            adapter.getItem(position)?.let { showItemActionDialog(it, rowView) }
            true
        }

        view.findViewById<FloatingActionButton>(R.id.addFAB).setOnClickListener {
            startActivity(Intent(requireContext(), AddItemActivity::class.java))
        }

        viewLifecycleOwner.lifecycleScope.launch {
            viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.uiState.collect { state -> adapter.submitList(state.items) }
            }
        }

        return view
    }

    private fun showItemActionDialog(item: StockedItem, rowView: View) {
        AlertDialog.Builder(requireContext())
            .setTitle(getString(R.string.item_action_title, item.product.name))
            .setMessage(R.string.item_action_message)
            .setPositiveButton(R.string.action_consumed) { _, _ ->
                viewModel.markConsumed(item.product)
            }
            .setNegativeButton(R.string.action_expired) { _, _ ->
                viewModel.markExpired(item.product)
            }
            .setNeutralButton(R.string.action_delete) { _, _ ->
                confirmDelete(item, rowView)
            }
            .show()
    }

    private fun confirmDelete(item: StockedItem, rowView: View) {
        AlertDialog.Builder(requireContext())
            .setTitle(R.string.delete_title)
            .setMessage(R.string.delete_message)
            .setPositiveButton(android.R.string.ok) { _, _ ->
                viewModel.delete(item.product)
                Snackbar.make(
                    rowView,
                    getString(R.string.deleted_confirmation, item.product.name),
                    Snackbar.LENGTH_LONG,
                ).show()
            }
            .setNegativeButton(android.R.string.cancel) { dialog, _ -> dialog.dismiss() }
            .show()
    }
}
